apache/airflow
# [AIRFLOW-149] Task Dependency Engine + Blocked Task Instance Explainer

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-149

Here is the original PR with Max's LGTM: https://github.com/aoen/incubator-airflow/pull/1
Since then I have made some fixes but this PR is essentially the same. It could definitely use more
eyes as there are likely still issues.

**Goals**
- Simplify, consolidate, and make consistent the logic of whether or not a task should be run
- Provide a view/better logging that gives insight into why a task instance is not currently running
  (no more viewing the scheduler logs to find out why a task instance isn't running for the majority
  of cases):
  ![image](https://cloud.githubusercontent.com/assets/1592778/17637621/aa669f5e-6099-11e6-81c2-d988d2073aac.png)

**Notable Functional Changes**
- Webserver view + task_failing_deps CLI command to explain why a given task instance isn't being
  run by the scheduler
- Running a backfill in the command line and running a task in the UI will now display detailed
  error messages based on which dependencies were not met for a task instead of appearing to succeed
  but actually failing silently
- Maximum task concurrency and pools are now respected by backfills
- Backfill now has the equivalent of the old force flag to run even for successful tasks
  This will break one use case:
  Using pools to restrict some resource on airflow executors themselves (rather than an externa
  l resource like a DB), e.g. some task uses 60% of cpu on a worker so we restrict that task's pool
  size to 1 to prevent two of the tasks from running on the same host. When backfilling a task of
  this type, now the backfill will wait on the pool to have slots open up before running the task
  even though we don't need to do this if backfilling on a different host outside of the pool. I
  think breaking this use case is OK since the use case is a hack due to not having a proper
  resource isolation solution (e.g. mesos should be used in this case instead).
- To make things less confusing for users, there is now a "ignore all dependencies" option for
  running tasks, "ignore dependencies" has been renamed to "ignore task dependencies", and "force"
  has been renamed to "ignore task instance state". The new "Ignore all dependencies" flag will
  ignore the following:
  - task instance's pool being full
  - execution date for a task instance being in the future
  - a task instance being in the retry waiting period
  - the task instance's task ending prior to the task instance's execution date
  - task instance is already queued
  - task instance has already completed
  - task instance is in the shutdown state
  - WILL NOT IGNORE task instance is already running
- SLA miss emails will now include all tasks that did not finish for a particular DAG run, even if
  the tasks didn't run because depends_on_past was not met for a task
- Tasks with pools won't get queued automatically the first time they reach a worker; if
  they are ready to run they will be run immediately
- Running a task via the UI or via the command line (backfill/run commands) will now log why a task
  could not get run if one if it's dependencies isn't met. For tasks kicked off via the web UI this
  means that tasks don't silently fail to get queued despite a successful message in the UI.
- Queuing a task into a pool that doesn't exist will now get stopped in the scheduler instead of a
  worker

**TODO**
- Fix test that is failing for mysql/postgres (not sure that it is related)

**Follow Up Items**
- Update the docs to reference the new explainer views/CLI command

**Testing Done**
- New unit tests
- Running in airbnb production for 2 days (will have been more by the time this is merged)
- Make sure blocked deps are all shown in UI, including dagrun doesn't exist dep
- Logs look good
- Make sure ui throws error when ti_deps aren't met when you try to schedule a task
- No delay/load issues
- Test celery worker starting a task via the UI with various flags (on Bronze), and make sure the
  task actually completed by checking start_date changed
- Run a task that has already succeeded in the UI and make sure it's state gets changed
  (including in the UI)
- Make sure can backfill overtop of failed tasks
- Make sure tasks still get queued when e.g. concurrency is reached or pool is reached (can test one
  don't have to test both)
- Make sure backfill flags still work (because they aren't passed to tasks anymore right?)
  -a, --include_adhoc Include dags with the adhoc parameter.
  -i, --ignore_dependencies
  Skip upstream tasks, run only the tasks matching the
  regexp. Only works in conjunction with task_regex
  -I, --ignore_first_depends_on_past
- Test airflow backfill and run commands with some flags
- Tasks that should have been deadlocked are still deadlocked after this change.
- Made sure that the right failing dependencies were shown on the task instance page.
- Tested new CLI command by running it
- Ui still works with all the various force flag combos (ignore_task_deps especially)
- Backfill before a task's start date should work
- Make sure that for the TI pop up dialog in tree view all buttons that are links work
- Make sure logs aren't too crazy (logging a lot more e.g. in scheduler)

**Future Work**
- Add ignore_ti_state flag to airflow backfill command to be consistent with UI/airflow run command.
- Show one more dagrun in the tree view (even when that dagrun's execution date hasn't occurred yet) so that users can see why task instances in the next dagrun are blocked. I wanted to include this in the first release but it's non-trivial to do (one way to solve this is to generate DAG runs before their execution date occurs). You can still view a non-existent task instance by passing the right query parameters to the task instance page, or by querying via the new command-line command.
- Break down failed dependency explanations better (e.g. if the trigger rules requiring all upstream task successes fails, indicate the specific upstream tasks that failed)
- Make failing dependencies for task instances more visible to users (e.g. asynchronously display failing dependencies when mousing over a task instance in the tree view).
- Parallelize task instance dependency checking.
- Don't pass around the ignore_x_deps type flags (e.g. ignore_all_deps), instead create a set of DependencyContext object at the first place the flags are initialized and pass this context down
- Can add some more tips to fix a failing dependency in the new view that shows failed task instance dependencies (e.g. for "depends_on_past" when the last task state is failing we can give the tip to get this task to succeed, or even provide a button to clear it right from that view)
- Additional task instance dependencies, e.g. DAG does not parse, there are no schedulers running (a task can't get scheduled until a scheduler is running), or the airflow scheduler queue is backed up, or no executors being available to run the task instance
- Keep history of changes to blocked dependencies that are accessible via the UI (this is useful to know why a task's execution was delayed)

@jlowin @mistercrunch @artwr 
cc: @shenghuy @PeterAttardo
